compiler/resolution/implementForallIntents.cpp

```diff
@@ -1819,88 +2102,46 @@ static void setShadowVarFlagsNew(Symbol* ovar, VarSymbol* svar, IntentTag intent
   //      plus perhaps "isReduce" and "pruneit" booleans
-  const int maxVars = uses.n;
+  const int maxVars = (int)fivec.size();
   outerVars.reserve(maxVars);
   shadowVars.reserve(maxVars);
-  reduceGVars.reserve(maxVars);
 
-  form_Map(SymbolMapElem, e, uses) {
-    if (e->value == markPruned)
-      continue;
+  // like for_vector()
+  for (FIIvec::iterator it = fivec.begin(); it != fivec.end(); it++) {
+    FIinfo& info = *it;
 
-    totOuterVars++;
-    Symbol* ovar = e->key;
+    INT_ASSERT(info.svar == NULL);
+
+    Symbol* ovar = info.ovar;
     // If ovar is a reference, e.g. an index variable of
     // a 'var' iterator, we do not want to force
     // that ref type onto 'svar'. Otherwise the generated
     // code will store into *svar without initializing svar
     // first. Todo: what if ovar is a domain?
     Type* valtype = ovar->type->getValType();
 
-    //
-    // Distinguish the following cases for the outer variable 'ovar':
-    //
-    //   (C1) e->value == markUnspecified
-    //        ==> ovar is not listed in the 'with' clause
-    //
-    //   (C2) e->value is one of "tiMark" ArgSymbols
-    //        ==> get the forall intent from that tiMark
-    //
-    //   (C3) e->value is a "chpl__reduceGlob" VarSymbol
-    //        ==> it is a reduce intent
-    //
-    // If (C1) or (C2), we may want to "prune" this variable,
-    // i.e. do not handle it under a forall intent, instead
-    // resort to the usual lexical scoping.
-    //
-    // If we do not prune, we compute (ovar, svar, globalOp)
-    // to push onto our lists.
-    //
-    bool isUnspec = (e->value == markUnspecified);
+    Symbol*    globalOp = info.reduceMark;
+    bool       isReduce = info.isReduce();
+    IntentTag  tiIntent = INTENT_BLANK; // not used for reduce intents
     VarSymbol* svar;
-    VarSymbol* globalOp = isUnspec ? NULL : toVarSymbol(e->value);
-    bool isReduce = globalOp != NULL;
-    IntentTag tiIntent = INTENT_BLANK;
 
     if (!isReduce) {
-      ArgSymbol* tiMarker = NULL;
-
-      // Compute the intent
-      if (isUnspec) {
-        // start with blank intent
-      } else {
-        tiMarker = toArgSymbol(e->value);
-        INT_ASSERT(tiMarker);
-        tiIntent = tiMarker->intent;
-      }
-
-      bool isMethodToken = ovar->type == dtMethodToken;
-
-      // concreteIntent() does not work for MT.
-      if (!isMethodToken)
-        tiIntent = concreteIntent(tiIntent, valtype);
+      tiIntent = concreteIntent(tfi2it(info.intent), valtype);
 
       // See if we want to prune it.
       bool pruneit = false;
 
       if (tiIntent == INTENT_REF || tiIntent == INTENT_REF_MAYBE_CONST) {
-        // for efficiency
-        pruneit = true;
-
-      } else if (isMethodToken) {
-        // If MT is present, _toLeader(..., _mt...) does not get resolved.
-        // See e.g. parallel/taskPar/figueroa/taskParallel.chpl
-        pruneit = true;
+        // do we want this? does this lead to more efficient generated code?
```

Comment:
This is where the question of forall intents vs. handling in flatten functions comes in. It's not obvious to me if it's better to leave this case to flatten functions but I lean a little bit towards having the forall intent code handle all captured variables. But, it's certainly the case that flatten functions is going to run anyway... Perhaps we want to outline these functions early? Perhaps forall intent processing is an early sort of flatten functions?

Anyway it might be less of a change to leave it this way.

Regarding the comment, I think it'd be worth directly saying that by pruning the outer variable reference, we leave it for flatten functions to handle. Unless that's wrong and I'm misunderstanding.